local-macros/src/lib.rs

```diff
@@ -1389,3 +1393,75 @@ pub fn symbol_type(attr: TokenStream, item: TokenStream) -> TokenStream {
     }
     .into()
 }
+
+fn get_enum_unwrapped_match(argument: &Expr, ancestors: &[Ident]) -> TokenStream2 {
+    let mut unwrapped_variant_selector = quote! {
+        unwrapped
+    };
+    for (index, ancestor) in ancestors.iter().skip(1).rev().enumerate() {
+        let previous_ancestor = &ancestors[ancestors.len() - 2 - index];
+        unwrapped_variant_selector = quote! {
+            crate::#previous_ancestor::#ancestor(#unwrapped_variant_selector)
+        }
+    }
+
+    let last_variant = ancestors.last().unwrap();
+
+    quote! {
+        match #argument {
+            #unwrapped_variant_selector => unwrapped,
+            _ => panic!("Expected {}", stringify!(#last_variant)),
+        }
+    }
+}
+
+fn expr_to_ident(expr: &Expr) -> Result<Ident> {
+    match expr {
+        Expr::Path(path) => match path.path.get_ident() {
+            Some(ident) => Ok(ident.clone()),
+            None => Err(Error::new_spanned(path, "Expected ident")),
+        },
+        _ => Err(Error::new_spanned(expr, "Expected ident")),
+    }
+}
+
+struct EnumUnwrapped {
+    argument: Expr,
+    ancestors: Vec<Ident>,
+}
+
+impl Parse for EnumUnwrapped {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let argument: Expr = input.parse()?;
+        input.parse::<Token![,]>()?;
```

Comment:
I was basing this off of an [example](https://github.com/dtolnay/syn/blob/master/examples/lazy-static/lazy-static/src/lib.rs) from `syn`, since it currently expects a known # of "arguments" this step-by-step parsing is pretty straightforward